File: batch-tech-feed/produce/techfeed/timeline.py
"""台本 + 音声の実尺からタイムラインを組み立てる。

このパイプラインには「手で打つタイムライン」が存在しない(docs/407 §2.3)。
スライドの切り替え時刻も字幕cueの時刻も、すべて合成された音声の実測尺から機械的に
導出する。ゆえに同じ台本から作り直せば常に同じ尺の動画になる(冪等)。
"""

from __future__ import annotations

from dataclasses import asdict, dataclass, field

from .script import Script, Section


@dataclass
class SlideState:
    """スライドの「ある瞬間の見た目」。1文につき1状態進む(docs/407 §2.4)。"""

    layout: str
    header: str
    title: str
    subtitle: str = ""
    items: list[str] = field(default_factory=list)
    # items のうち何件目までを表示するか。箇条書きを1項目ずつ出すために使う。
    revealed: int = 0
    # 現在読み上げている項目のindex。ハイライト表示に使う。
    highlight: int | None = None
    code: str = ""
    language: str = ""
    image: str = ""
    caption: str = ""
    diagram: str = ""
    sources: list[str] = field(default_factory=list)
    # セクション番号/総数。進捗インジケータに使う。
    section_seq: int = 1
    section_total: int = 1



@dataclass
class Entry:
    """ナレーション1文に対応する、音声・スライド・字幕の一体となった単位。"""

    seq: int
    section_seq: int
    text: str
    state: SlideState
    # 以下は音声合成が終わって初めて確定する。
    audio_path: str = ""
    start_ms: int = 0
    end_ms: int = 0

    @property
    def duration_ms(self) -> int:
        return self.end_ms - self.start_ms


def _states_for_section(section: Section, header: str, total: int) -> list[SlideState]:
    """セクション内の各文に対応するスライド状態を作る。"""
    slide = section.slide
    sources = [s.publisher or s.url for s in section.sources if (s.publisher or s.url)]

    base = SlideState(
        layout=slide.layout,
        header=header,
        title=slide.title,
        subtitle=slide.subtitle,
        items=list(slide.items),
        code=slide.code,
        language=slide.language,
        image=slide.image,
        caption=slide.caption,
        diagram=slide.diagram,
        sources=sources,
        section_seq=section.seq,
        section_total=total,
    )

    states: list[SlideState] = []
    revealed = 0
    highlight: int | None = None
    for line in section.narration:
        if slide.layout == "bullets" and line.focus is not None:
            # 既に出した項目は引っ込めない。focus は「ここまで出して、ここを強調」を意味する。
            revealed = max(revealed, line.focus + 1)
            highlight = line.focus
        # focus が None のときは状態を据え置く。1項目を複数文で語れるようにするため。
        states.append(
            SlideState(
                **{**asdict(base), "revealed": revealed, "highlight": highlight}
            )
        )
    return states


def build(script: Script) -> list[Entry]:
    """台本からタイムラインの骨格を作る。この時点では尺は未確定(すべて0)。"""
    total = len(script.sections)
    header = script.title

    entries: list[Entry] = []
    seq = 0
    for section in script.sections:
        states = _states_for_section(section, header, total)
        for line, state in zip(section.narration, states):
            entries.append(
                Entry(seq=seq, section_seq=section.seq, text=line.text, state=state)
            )
            seq += 1
    return entries


def apply_durations(entries: list[Entry], durations_ms: list[int]) -> int:
    """実測した各文の尺を積み上げて start/end を確定させ、総尺(ms)を返す。

    尺は「計算」ではなく ffprobe による「実測」を渡すこと。想定尺と実ファイルの尺が
    ずれると、動画が進むにつれて字幕とスライドが少しずつ後ろへずれていくため。
    """
    if len(entries) != len(durations_ms):
        raise ValueError(
            f"entry/duration count mismatch: {len(entries)} entries, {len(durations_ms)} durations"
        )

    cursor = 0
    for entry, dur in zip(entries, durations_ms):
        entry.start_ms = cursor
        cursor += dur
        entry.end_ms = cursor
    return cursor


def to_manifest(
    script: Script, entries: list[Entry], fps: int, narration_path: str = ""
) -> dict:
    """レンダラに依存しない、動画1本の完全な記述。

    このパイプラインの核は「音声の実尺からタイムラインを導く」ことであり、それは
    どのレンダラを使うかとは独立している(docs/407 §2.4)。よって manifest を
    Python 側と描画側の唯一の契約とし、スライドの状態までここに書き出す。
    描画側が focus の解釈をやり直さずに済み、字幕の投入(MVP-2)も同じ物を読める。
    """
    return {
        "digest_date": script.digest_date,
        "title": script.title,
        "fps": fps,
        "total_ms": entries[-1].end_ms if entries else 0,
        "narration": narration_path,
        "entries": [
            {
                "seq": e.seq,
                "section_seq": e.section_seq,
                "text": e.text,
                "start_ms": e.start_ms,
                "end_ms": e.end_ms,
                "audio": e.audio_path,
                "state": asdict(e.state),
            }
            for e in entries
        ],
    }

File: batch-tech-feed/produce/shared/timeline.go
package shared

import (
	"encoding/json"
	"fmt"
	"os"
)

// SlideState はスライドのある瞬間の状態。
type SlideState struct {
	Layout       string   `json:"layout"`
	Header       string   `json:"header"`
	Title        string   `json:"title"`
	Subtitle     string   `json:"subtitle"`
	Items        []string `json:"items"`
	Revealed     int      `json:"revealed"`
	Highlight    *int     `json:"highlight"`
	Code         string   `json:"code"`
	Language     string   `json:"language"`
	Image        string   `json:"image"`
	Caption      string   `json:"caption"`
	Diagram      string   `json:"diagram"`
	Sources      []string `json:"sources"`
	SectionSeq   int      `json:"section_seq"`
	SectionTotal int      `json:"section_total"`
}

// Entry はナレーション1文に対応する音声・スライド・字幕の単位。
type Entry struct {
	Seq        int        `json:"seq"`
	SectionSeq int        `json:"section_seq"`
	Text       string     `json:"text"`
	State      SlideState `json:"state"`
	AudioPath  string     `json:"audio"`
	StartMs    int        `json:"start_ms"`
	EndMs      int        `json:"end_ms"`
}

func (e *Entry) DurationMs() int {
	return e.EndMs - e.StartMs
}

func BuildTimeline(s *Script) []*Entry {
	total := len(s.Sections)
	header := s.Title

	var entries []*Entry
	seq := 0
	for _, sec := range s.Sections {
		states := statesForSection(sec, header, total)
		for i, line := range sec.Narration {
			state := states[i]
			entries = append(entries, &Entry{
				Seq:        seq,
				SectionSeq: sec.Seq,
				Text:       line.Text,
				State:      state,
			})
			seq++
		}
	}
	return entries
}

func statesForSection(sec Section, header string, total int) []SlideState {
	slide := sec.Slide
	var sources []string
	for _, src := range sec.Sources {
		val := src.Publisher
		if val == "" {
			val = src.URL
		}
		if val != "" {
			sources = append(sources, val)
		}
	}
	if sources == nil {
		sources = []string{}
	}

	items := slide.Items
	if items == nil {
		items = []string{}
	}

	base := SlideState{
		Layout:       slide.Layout,
		Header:       header,
		Title:        slide.Title,
		Subtitle:     slide.Subtitle,
		Items:        items,
		Code:         slide.Code,
		Language:     slide.Language,
		Image:        "",
		Caption:      "",
		Diagram:      slide.Diagram,
		Sources:      sources,
		SectionSeq:   sec.Seq,
		SectionTotal: total,
	}

	var states []SlideState
	revealed := 0
	for _, line := range sec.Narration {
		var highlight *int
		if slide.Layout == "bullets" && line.Focus != nil {
			idx := *line.Focus
			if idx+1 > revealed {
				revealed = idx + 1
			}
			highlight = line.Focus
		}

		st := base
		st.Revealed = revealed
		st.Highlight = highlight
		states = append(states, st)
	}

	return states
}

func ApplyDurations(entries []*Entry, durationsMs []int) (int, error) {
	if len(entries) != len(durationsMs) {
		return 0, fmt.Errorf("entry/duration count mismatch: %d entries, %d durations", len(entries), len(durationsMs))
	}

	cursor := 0
	for i, entry := range entries {
		dur := durationsMs[i]
		entry.StartMs = cursor
		cursor += dur
		entry.EndMs = cursor
	}
	return cursor, nil
}

type Manifest struct {
	DigestDate string   `json:"digest_date"`
	Title      string   `json:"title"`
	FPS        int      `json:"fps"`
	TotalMs    int      `json:"total_ms"`
	Narration  string   `json:"narration"`
	Entries    []*Entry `json:"entries"`
}

func ToManifest(s *Script, entries []*Entry, fps int, narrationPath string) *Manifest {
	totalMs := 0
	if len(entries) > 0 {
		totalMs = entries[len(entries)-1].EndMs
	}

	return &Manifest{
		DigestDate: s.DigestDate,
		Title:      s.Title,
		FPS:        fps,
		TotalMs:    totalMs,
		Narration:  narrationPath,
		Entries:    entries,
	}
}

func SaveManifest(m *Manifest, path string) error {
	data, err := json.MarshalIndent(m, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(path, data, 0o644)
}
